### PR TITLE
Fact samplers

### PR DESCRIPTION
**Pull request by mschauer**

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -22,3 +22,48 @@ struct Boomerang1d{T} <: ContinuousDynamics
...
+target precision. μ is the approximate target mean.
+"""
+struct ZigZag{T,S} <: ContinuousDynamics
+    Γ::T
+    μ::S
+end
+
+"""
+    Bps{T} <: ContinuousDynamics
+λref::T : refreshment rate which has to be strivtly positive
```
Comment (mschauer) [inline src/types.jl:40]:
```suggestion
λref::T : refreshment rate which has to be strictly positive
```

File: src/fact_samplers.jl [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    b = 0.0
+    a, b
+end
+
+"""
+    λ_bar(G, i, x, θ, c, Z::ZigZag)
+
+Computes the bounding rate `λ_bar` at position `x` and velocity `θ`.
+"""
+λ_bar(G, i, x, θ, c, Z::ZigZag) = pos(ab(G, i, x, θ, c, Z::ZigZag)[1])
```
Comment (mschauer) [inline src/fact_samplers.jl:77]:
```suggestion
λ_bar(G, i, x, θ, c, Z::ZigZag) = pos(ab(G, i, x, θ, c, Z)[1])
```

File: src/fact_samplers.jl [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+"""
+λ_bar(G, i, x, θ, c, Z::FactBoomerang) = pos(ab(G, i, x, θ, c, Z)[1])
+
+
+function event(i, t, x, θ, Z::ZigZag)
+     t, i, x[i], θ[i]
+end
+
+function event(i, t, x, θ, Z::FactBoomerang)
+    t, i, x, θ
```
Comment (mschauer) [inline src/fact_samplers.jl:92]: Not `x[i]` and `θ[i]`?

Comment (SebaGraz) [reply src/fact_samplers.jl:92]: Ah! You are right! `x[i]` and `θ[i]` are sufficient.

File: src/trace.jl [hunk 4]
```diff
@@ -33,7 +66,7 @@ end
 Base.IteratorSize(::Discretize) = Iterators.SizeUnknown()
 discretize(Z, dt) = Discretize(Z, dt)
 
-function Base.iterate(D::Discretize{<:ZigZagTrace})
+function Base.iterate(D::Discretize{<:Union{ZigZagTrace, FactBoomTrace}})
```
Comment (mschauer) [inline src/trace.jl:69]:
```suggestion
function Base.iterate(D::Discretize{T}) where {T <: Union{ZigZagTrace, FactBoomTrace}}
```

**Pull request closed without merging**